peps/pep-0750.rst

```diff
@@ -0,0 +1,813 @@
+
+.. code-block:: python
+
+    mytag()
+
+HTML Example of Rich Return Types
+=================================
+
+Tag functions can be a powerful part of larger processing chains by returning richer objects.
+JavaScript tagged template literals, for example, are not constrained by a requirement to
+return a string. As an example, let's look at an HTML generation system, with a usage and
+"subcomponent":
+
+.. code-block::
+
+    def Menu(*, logo: str, class_: str) -> HTML:
+        return html'<img alt="Site Logo" src={logo} class={class_} />'
+
+    icon = 'acme.png'
+    result = html'<header><{Menu} logo={icon} class="my-menu"/></header>'
+    img = result.children[0]
+    assert img.tag == "img"
+    assert img.attrs == {"src": "acme.png", "class": "my-menu", "alt": "Site Logo"}
+    # We can also treat the return type as a string of specially-serialized HTML
+    assert str(result) = '<header>' # etc.
+
+This ``html`` tag function might have the following signature:
+
+.. code-block:: python
+
+    def html(*args: Decoded | Interpolation) -> HTML:
+        ...
+
+The ``HTML`` return class might have the following shape as a ``Protocol``:
+
+.. code-block:: python
+
+    @runtime_checkable
+    class HTML(Protocol):
+        tag: str
+        attrs: dict[str, Any]
+        children: Sequence[str | HTML]
+
+In summary, the returned instance can be used as:
+
+- A string, for serializing to the final output
+- An iterable, for working with WSGI/ASGI for output streamed and evaluated
+  interpolations *in the order* they are written out
+- A DOM (data) structure of nested Python data
+
+In each case, the result can be lazily and recursively composed in a safe fashion, because
+the return value isn't required to be a string. Recommended practice is that
+return values are "passive" objects.
+
+What benefits might come from returning rich objects instead of strings? A DSL for
+a domain such as HTML templating can provide a toolchain of post-processing, as
+`Babel <https://babeljs.io>`_ does for JavaScript
+`with AST-based transformation plugins <https://babeljs.io/docs/#pluggable>`_.
+Similarly, systems that provide middleware processing can operate on richer,
+standard objects with more capabilities. Tag string results can be tested as
+nested Python objects, rather than string manipulation. Finally, the intermediate
+results can be cached/persisted in useful ways.
+
+Tool Support
+============
+
+Python Semantics in Tag Strings
+-------------------------------
+
+Python template languages and other DSLs have semantics quite apart from Python.
+Different scope rules, different calling semantics e.g. for macros, their own
+grammar for loops, and the like.
+
+This means all tools need to write special support for each language. Even then,
+it is usually difficult to find all the possible scopes, for example to autocomplete
+values.
+
+f-strings of course do not have this issue. An f-string is considered part of Python.
```

Comment:
```suggestion
F-strings of course do not have this issue. An f-string is considered part of Python.
```

Reply:
Let's rewrite this to avoid the upper casing of "f" in f-string. Something like
> However, f-strings do not have this issue...